plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/resources/OpenshiftResourceChangeListener.java

```diff
@@ -0,0 +1,254 @@
+						return false;
+					}
+				}
+				// Still here means delta has changes
+				changed.add(delta2.getResource());
+				return true;
+			}
+		};
+		try {
+			delta.accept(visitor);
+		} catch(CoreException ce) {
+			// TODO log
+		}
+		return changed;
+	}
+	
+	protected void publishHandleProjectChange(IResourceDelta delta, IResourceChangeEvent event) {
+		IProject project = (IProject) delta.getResource();
+		if (project == null)
+			return;
+
+		List<IResource> changes = getDeltaResourceChanges(delta); 
+		if( changes.size() > 0) {
+			OpenShiftServer[] servers2 = getPublishRequiredServers(delta);
+			int size2 = servers2.length;
+			for (int j = 0; j < size2; j++) {
+				handleSpecialProjectChange(servers2[j], delta, changes);
+			}
+		}
+	}
+	
+	private OpenShiftServer[] getPublishRequiredServers(IResourceDelta delta){		
+		// The list of servers that will require publish
+		final List<OpenShiftServer> servers2 = new ArrayList<>();
+
+		// wrksServers = Workspaces Servers
+		final IServer[] wrksServers =  ServerCore.getServers();
+		for( int i = 0; i < wrksServers.length; i++ ) {
+			OpenShiftServer os = (OpenShiftServer)wrksServers[i].loadAdapter(OpenShiftServer.class, new NullProgressMonitor());
+			if( os != null ) {
+				IProject magic = OpenShiftServerUtils.getDeployProject(wrksServers[i]);
+				if( magic != null ) {
+					 // Safe because we've already eliminated non-project deltas
+					IProject p = (IProject)delta.getResource();
+					if( magic.equals(p)) {
+						servers2.add(os);
+					}
+				}
+			}
+		}
+		return (OpenShiftServer[]) servers2.toArray(new OpenShiftServer[servers2.size()]);
+	}
+	
+
+	protected void handleSpecialProjectChange(OpenShiftServer server, IResourceDelta delta,  List<IResource> changes) {
+		// check for duplicate jobs already waiting and don't create a new one
+		Job[] jobs = Job.getJobManager().find(ServerUtil.SERVER_JOB_FAMILY);
+		if (jobs != null) {
+			int size = jobs.length;
+			for (int i = 0; i < size; i++) {
+				if (jobs[i] instanceof MagicProjectChangeJob) {
+					MagicProjectChangeJob rcj = (MagicProjectChangeJob) jobs[i];
+					if (rcj.getServer().equals(server.getServer()) && rcj.getState() == Job.WAITING)
+						return;
+				}
+			}
+		}
+		
+		MagicProjectChangeJob job = new MagicProjectChangeJob(server, delta, changes);
+		job.setSystem(true);
+		job.setPriority(Job.BUILD);
+		job.schedule();
+	}
+	
+	/**
+	 * Give the server an indication that its magic project has been changed somehow
+	 */
+	public class MagicProjectChangeJob extends Job {
+		private IResourceDelta delta;
+		private OpenShiftServer openshiftServer;
+		private IResourceChangeEvent event;
+		private List<IResource> changes;
+
+		public MagicProjectChangeJob(OpenShiftServer openshiftServer, IResourceDelta delta, List<IResource> changes) {
+			super(NLS.bind(Messages.jobUpdateServer, openshiftServer.getServer().getName()));
+			this.openshiftServer = openshiftServer;
+			this.delta = delta;
```

Comment:
IResourceDelta passed to MagicProjectChangeJob is never used which is good because as docs warn the resource delta object 'is valid only for the duration of the invocation of' resourceChanged(IResourceChangeEvent), it may not be valid when job is runned. I suggest removing this parameter.